paper.md

```diff
@@ -120,15 +120,63 @@ Several hosted CI services like Travis, Circle and Jenkins make continuous integ
+
+A popper workflow consists of a series of syntactical components called steps, where each step represents a node in the workflow DAG, with a `uses` attribute specifying the required container image. 
+The `uses` attribute can reference Docker images hosted in container image registries; filesystem paths for locally defined container images (Dockerfiles); or publicly accessible GitHub repositories that contain Dockerfiles. 
+The commands or scripts that need to be executed in a container can be defined by the `args` and `runs` attributes. 
+Secrets and environment variables needed by a step can be specified by the `secrets` and `env` attributes respectively for making them available inside the container associated to a step.
+The steps in a workflow are executed sequentially in the order in which they are defined.
+
 ## Workflow Execution Engine
 
+The Popper workflow execution engine is composed of several components which talk to each other during a workflow execution.
+The vital architectural components of the system are described in detail throughout this section.
+
 ### **Command line interface (CLI)**
 
-### **Workflow Runner**
+Besides allowing users to communicate with the workflow runner, the CLI provides with search capabilities so that users can search for existing actions implemented by other researchers and developers;
+allows visualizing workflows by generating DOT diagrams [@dot] like the one shown in @Fig:casestudy;
+generates configuration files for continuous integration systems, e.g. TravisCI, Jenkins, Gitlab-CI, etc. so that users can continuously validate the workflows they implement;
+provides dynamic workflow variable substitution capabilities, among others.
 
-### **Resource manager API**
+### **Workflow Runner**
 
-### **Container engine API and plugins**
+The Workflow runner is in charge of parsing the `.yml` files, i.e. the workflow files and the configuration files through their corresponding parsers and creating an internal representation for it. 
+It also downloads actions referenced by the steps in the workflow, checks the presence of secrets that are used in a workflow and routes the execution of a step to the configured container engine through the requested resource manager. 
+The runner also maintains a cache directory to optimize multiple aspects of execution such as avoid cloning repositories if they have already cloned previously. 
+Thus, this component orchestrates the entire workflow execution process.
+
+### **Parser**
+
+The workflow file and the configuration file are parsed by their respective parser plugins at the initial stages of a workflow execution.
+The parsers are responsible for reading and parsing the YML files into an internal format;
+running syntactic and semantic validation checks;
+normalizing the various attributes and generating a workflow DAG.
+The parsed file contents are represented internally with the help of different Python data structures like `dict` and `list`.
```

Comment:
this sentence might be too low-level, so we might be able to remove it

Reply:
@ivotron There was a comment in the review to describe the internal format of the workflow briefly. Can you give some idea on how to frame it ?

Reply:
This is still left.